File: c_nif/native/lib/calc.c
#include <stdlib.h>
#include "calc.h"

int add_int(int a, int b)
{
    return a + b;
}

int div_int(int a, int b)
{
    return a / b;
}

double div_double(double a, double b)
{
    return a / b;
}

void panic()
{
    exit(1);
}

File: c_nif/lib/calc.ex
defmodule Calc do
  @on_load :load_nifs

  def load_nifs do
    :erlang.load_nif('native/nif_calc', 0)
  end

  def add_int(_a, _b), do: :erlang.error(:nif_not_loaded)

  def div_int(_a, _b), do: :erlang.error(:nif_not_loaded)

  def div_double(_a, _b), do: :erlang.error(:nif_not_loaded)

  def panic(), do: :erlang.error(:nif_not_loaded)
end
